llvm/lib/Transforms/Scalar/LICM.cpp

```diff
@@ -2850,7 +2855,7 @@ static bool hoistBOAssociation(Instruction &I, Loop &L,
   bool LVInRHS = L.isLoopInvariant(BO->getOperand(0));
   auto *BO0 = dyn_cast<BinaryOperator>(BO->getOperand(LVInRHS));
   if (!BO0 || BO0->getOpcode() != Opcode || !BO0->isAssociative() ||
-      BO0->hasNUsesOrMore(3))
+      BO0->hasNUsesOrMore(BOAssociationUserLimit + 1))
```

Comment:
Hm, isn't this actually a lower-bound (N Uses Or More)? Your description "Limit the number of users ..." doesn't seem to align. Also, what is the motivation behind this change?

Reply:
`hasNUsesOrMore(N+1)` itself is indeed a lower bound. However, in this case, if it's true then (next line) we don't perform the re-association. We are essentially checking if the number of uses is at most `N`. 

I could rewrite this to `!BO0->getNumUses() <= BOAssociationUserLimit` to make it clearer.

Reply:
Regarding the motivation: we (at NVIDIA) have found that in some cases too much re-association can lead to higher register pressure, especially when there are multiple uses inside a loop. We'd like a way to control this behavior.

Reply:
Is LICM the correct place to fix this though? LICM is some form of canonicalization, w/o any consideration for register pressure. Is there any reason the backend cannot sink the instructions back if beneficial?

Reply:
In principle it makes sense to fix such RP issues in the backend. However, in our case, the the re-associated and hoisted instructions are further modified by other passes (LSR etc) and sinking them becomes non-trivial.